Add miri support

Discussed briefly on IRC with @shepmaster: It would be great if the playground could run programs in [miri](https://github.com/solson/miri). That's a MIR interpreter. One reason why one might want to do this is that miri will stop program execution when there is UB -- the checks are not complete yet, but more are coming. It will never be able to detect *all* kinds of UB (e.g. relying on how `#[repr(Rust)]` lays out data is hard to detect), but the idea is it should detect e.g. violation of the pointer aliasing rules.

Since I don't know how the playground works, I'll just give instructions for how to run miri manually. All of this has to happen with a nightly compiler.

Setup:

* `git clone https://github.com/solson/miri && cd miri`
* `cargo install xargo`
* `xargo/build.sh`
* Now xargo is no longer needed and could be uninstalled; the purpose was just to obtain the files that now live in `~/.xargo/HOST`.
* `cargo install --all-features --force`

Now, when you are in a crate, instead of `cargo run` you would do `MIRI_SYSROOT=~/.xargo/HOST cargo miri -- -Zmiri-start-fn`. That's it.

> improve the error message to make it easier to interpret

In that realm, the output is really a wall of text, very hard to read. I think if you want people to easily drop in and start using Miri, you might need to pay some mind to that:

```
   Compiling playground v0.0.1 (file:///playground)
error[E0080]: constant evaluation error: pointer computed at offset 1, outside bounds of allocation 416 which has size 0
    --> /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/ptr.rs:1383:9
     |
1383 |         intrinsics::offset(self, count) as *mut T
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ pointer computed at offset 1, outside bounds of allocation 416 which has size 0
     |
note: inside call to `std::ptr::<impl *mut T><u8>::offset`
    --> /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/slice/mod.rs:2079:15
     |
2079 |         &mut *slice.as_mut_ptr().offset(self as isize)
     |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside call to `<usize as std::slice::SliceIndex<[T]>><u8>::get_unchecked_mut`
    --> /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/slice/mod.rs:405:9
     |
405  |         index.get_unchecked_mut(self)
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside call to `core::slice::<impl [T]><u8>::get_unchecked_mut::<usize>`
    --> src/main.rs:3:15
     |
3    |     unsafe { *a.get_unchecked_mut(1) = 1; }
     |               ^^^^^^^^^^^^^^^^^^^^^^
note: inside call to `main`
    --> /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:74:34
     |
74   |     lang_start_internal(&move || main().report(), argc, argv)
     |                                  ^^^^^^
note: inside call to `closure`
    --> /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:59:75
     |
59   |             ::sys_common::backtrace::__rust_begin_short_backtrace(move || main())
     |                                                                           ^^^^^^
note: inside call to `closure`
    --> /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/sys_common/backtrace.rs:136:5
     |
136  |     f()
     |     ^^^
note: inside call to `std::sys_common::backtrace::__rust_begin_short_backtrace::<[closure@DefId(1/1:1839 ~ std[82ff]::rt[0]::lang_start_internal[0]::{{closure}}[0]::{{closure}}[0]) 0:&dyn std::ops::Fn() -> i32 + std::panic::RefUnwindSafe + std::marker::Sync], i32>`
    --> /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:59:13
     |
59   |             ::sys_common::backtrace::__rust_begin_short_backtrace(move || main())
     |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside call to `closure`
    --> /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:310:40
     |
310  |             ptr::write(&mut (*data).r, f());
     |                                        ^^^
note: inside call to `std::panicking::try::do_call::<[closure@DefId(1/1:1838 ~ std[82ff]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&&dyn std::ops::Fn() -> i32 + std::panic::RefUnwindSafe + std::marker::Sync], i32>`
    --> /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:306:5
     |
306  | /     fn do_call<F: FnOnce() -> R, R>(data: *mut u8) {
307  | |         unsafe {
308  | |             let data = data as *mut Data<F, R>;
309  | |             let f = ptr::read(&mut (*data).f);
310  | |             ptr::write(&mut (*data).r, f());
311  | |         }
312  | |     }
     | |_____^
note: inside call to `std::panicking::try::<i32, [closure@DefId(1/1:1838 ~ std[82ff]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&&dyn std::ops::Fn() -> i32 + std::panic::RefUnwindSafe + std::marker::Sync]>`
    --> /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panic.rs:392:9
     |
392  |         panicking::try(f)
     |         ^^^^^^^^^^^^^^^^^
note: inside call to `std::panic::catch_unwind::<[closure@DefId(1/1:1838 ~ std[82ff]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&&dyn std::ops::Fn() -> i32 + std::panic::RefUnwindSafe + std::marker::Sync], i32>`
    --> /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:58:25
     |
58   |           let exit_code = panic::catch_unwind(|| {
     |  _________________________^
59   | |             ::sys_common::backtrace::__rust_begin_short_backtrace(move || main())
60   | |         });
     | |__________^
note: inside call to `std::rt::lang_start_internal`
    --> /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:74:5
     |
74   |     lang_start_internal(&move || main().report(), argc, argv)
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```

> E.g. we could try to not show spans in libstd/libcore? Or any other crate than the current one? Not sure.

Heh, not installing rust-src gives what seems like reasonable output:

```
error[E0080]: constant evaluation error: pointer computed at offset 1, outside bounds of allocation 416 which has size 0
     |
note: inside call to `std::ptr::<impl *mut T><u8>::offset`
note: inside call to `<usize as std::slice::SliceIndex<[T]>><u8>::get_unchecked_mut`
note: inside call to `core::slice::<impl [T]><u8>::get_unchecked_mut::<usize>`
    --> src/main.rs:6:10
     |
6    |         *a.get_unchecked_mut(1) = 1;
     |          ^^^^^^^^^^^^^^^^^^^^^^
note: inside call to `main`
note: inside call to `closure`
note: inside call to `closure`
note: inside call to `std::sys_common::backtrace::__rust_begin_short_backtrace::<[closure@DefId(1/1:1839 ~ std[82ff]::rt[0]::lang_start_internal[0]::{{closure}}[0]::{{closure}}[0]) 0:&dyn std::ops::Fn() -> i32 + std::panic::RefUnwindSafe + std::marker::Sync], i32>`
note: inside call to `closure`
note: inside call to `std::panicking::try::do_call::<[closure@DefId(1/1:1838 ~ std[82ff]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&&dyn std::ops::Fn() -> i32 + std::panic::RefUnwindSafe + std::marker::Sync], i32>`
note: inside call to `std::panicking::try::<i32, [closure@DefId(1/1:1838 ~ std[82ff]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&&dyn std::ops::Fn() -> i32 + std::panic::RefUnwindSafe + std::marker::Sync]>`
note: inside call to `std::panic::catch_unwind::<[closure@DefId(1/1:1838 ~ std[82ff]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&&dyn std::ops::Fn() -> i32 + std::panic::RefUnwindSafe + std::marker::Sync], i32>`
note: inside call to `std::rt::lang_start_internal`
```

I wonder if "current crate" + "bottommost stack frame" would be useful.

--- a/.travis.yml
+++ b/.travis.yml
@@ -38,6 +38,9 @@ jobs:
     - env:
         - TOOLS_TO_BUILD="clippy"
       script: ./.travis/build-containers.sh || true
+    - env:
+        - TOOLS_TO_BUILD="miri"
+      script: ./.travis/build-containers.sh || true
 
     - stage: compile code
       env:

--- a/compiler/build.sh
+++ b/compiler/build.sh
@@ -3,7 +3,7 @@
 set -euv -o pipefail
 
 channels_to_build="${CHANNELS_TO_BUILD-stable beta nightly}"
-tools_to_build="${TOOLS_TO_BUILD-rustfmt clippy}"
+tools_to_build="${TOOLS_TO_BUILD-rustfmt clippy miri}"
 perform_push="${PERFORM_PUSH-false}"
 
 repository=shepmaster

--- a/compiler/fetch.sh
+++ b/compiler/fetch.sh
@@ -4,8 +4,8 @@ set -euv -o pipefail
 
 repository=shepmaster
 
-for image in rust-stable rust-beta rust-nightly rustfmt clippy; do
+for image in rust-stable rust-beta rust-nightly rustfmt clippy miri; do
     docker pull "${repository}/${image}"
-    # The backend expects images without a respoitory prefix
+    # The backend expects images without a repository prefix
     docker tag "${repository}/${image}" "${image}"
 done

--- a/compiler/miri/Dockerfile
+++ b/compiler/miri/Dockerfile
@@ -1,0 +1,17 @@
+FROM shepmaster/rust-nightly as sources
+
+RUN rustup component add rust-src
+
+FROM sources as build
+
+RUN cargo install xargo
+RUN git clone https://github.com/solson/miri ~/miri
+RUN ~/miri/xargo/build.sh
+RUN cargo install --all-features --path ~/miri
+
+FROM sources
+
+COPY --from=build /root/.xargo /root/.xargo
+COPY --from=build /root/.cargo/bin/miri /root/.cargo/bin
+COPY --from=build /root/.cargo/bin/cargo-miri /root/.cargo/bin
+ADD cargo-miri-playground /root/.cargo/bin

--- a/compiler/miri/cargo-miri-playground
+++ b/compiler/miri/cargo-miri-playground
@@ -1,0 +1,6 @@
+#!/usr/bin/env bash
+
+set -eu
+
+export MIRI_SYSROOT=~/.xargo/HOST
+exec cargo miri -- -Zmiri-start-fn

--- a/tests/spec/features/tools_spec.rb
+++ b/tests/spec/features/tools_spec.rb
@@ -37,6 +37,24 @@ RSpec.feature "Using third-party Rust tools", type: :feature, js: true do
     EOF
   end
 
+  scenario "sanitize code with Miri" do
+    editor.set code_with_undefined_behavior
+    in_tools_menu { click_on("Miri") }
+
+    within(".output-stderr") do
+      expect(page).to have_content /pointer computed at offset 1, outside bounds of allocation \d+ which has size 0/
+    end
+  end
+
+  def code_with_undefined_behavior
+    <<~EOF
+    fn main() {
+        let mut a: [u8; 0] = [];
+        unsafe { *a.get_unchecked_mut(1) = 1; }
+    }
+    EOF
+  end
+
   def editor
     Editor.new(page)
   end

--- a/ui/frontend/Help.tsx
+++ b/ui/frontend/Help.tsx
@@ -9,6 +9,7 @@ import integer32Logo from './assets/integer32-logo.svg';
 
 const ACE_URL = 'https://github.com/ajaxorg/ace';
 const CLIPPY_URL = 'https://github.com/Manishearth/rust-clippy';
+const MIRI_URL = 'https://github.com/solson/miri';
 const CRATES_IO_URL = 'https://crates.io/';
 const RUST_COOKBOOK_URL = 'https://rust-lang-nursery.github.io/rust-cookbook/';
 const CRATES_URL = 'https://github.com/integer32llc/rust-playground/blob/master/compiler/base/Cargo.toml';
@@ -32,6 +33,13 @@ const CLIPPY_EXAMPLE = `fn main() {
     match true {
         true => println!("true"),
         false => println!("false"),
+    }
+}`;
+
+const MIRI_EXAMPLE = `fn main() {
+    let mut a: [u8; 0] = [];
+    unsafe {
+        *a.get_unchecked_mut(1) = 1;
     }
 }`;
 
@@ -137,6 +145,17 @@ const Help = ({ navigateToIndex }: HelpProps) => (
         </p>
 
         <Example code={CLIPPY_EXAMPLE} />
+      </LinkableSection>
+
+      <LinkableSection id="features-miri" header="Checking code for undefined behavior" level={H3}>
+        <p>
+          <a href={MIRI_URL}>Miri</a> is an interpreter for Rust's mid-level intermediate
+          representation (MIR) and can be used to detect certain kinds of undefined behavior
+          in your unsafe Rust code. Click on the <strong>Miri</strong> button in
+          the <strong>Tools</strong> menu to check.
+        </p>
+
+        <Example code={MIRI_EXAMPLE} />
       </LinkableSection>
 
       <LinkableSection id="features-sharing" header="Sharing code" level={H3}>

--- a/ui/frontend/Help.tsx
+++ b/ui/frontend/Help.tsx
@@ -116,6 +116,17 @@ const Help = ({ navigateToIndex }: HelpProps) => (
         </p>
       </LinkableSection>
 
+      <LinkableSection id="features-formatting" header="Formatting code" level={H3}>
+        <p>
+          <a href={RUSTFMT_URL}>rustfmt</a> is a tool for formatting Rust code
+          according to the Rust style guidelines. Click on the <strong>Format</strong>
+          {' '}
+          button in the <strong>Tools</strong> menu to automatically reformat your code.
+        </p>
+
+        <Example code={RUSTFMT_EXAMPLE} />
+      </LinkableSection>
+
       <LinkableSection id="features-linting" header="Linting code" level={H3}>
         <p>
           <a href={CLIPPY_URL}>Clippy</a> is a collection of lints to catch common
@@ -126,17 +137,6 @@ const Help = ({ navigateToIndex }: HelpProps) => (
         </p>
 
         <Example code={CLIPPY_EXAMPLE} />
-      </LinkableSection>
-
-      <LinkableSection id="features-formatting" header="Formatting code" level={H3}>
-        <p>
-          <a href={RUSTFMT_URL}>rustfmt</a> is a tool for formatting Rust code
-          according to the Rust style guidelines. Click on the <strong>Format</strong>
-          {' '}
-          button in the <strong>Tools</strong> menu to automatically reformat your code.
-        </p>
-
-        <Example code={RUSTFMT_EXAMPLE} />
       </LinkableSection>
 
       <LinkableSection id="features-sharing" header="Sharing code" level={H3}>

--- a/ui/frontend/Output.tsx
+++ b/ui/frontend/Output.tsx
@@ -120,7 +120,7 @@ interface FormatProps {
 const Output: React.SFC<OutputProps> = ({
   // https://github.com/palantir/tslint/issues/3960
   // tslint:disable-next-line:trailing-comma
-  somethingToShow, meta: { focus }, execute, format, clippy, assembly, llvmIr, mir, wasm, gist, ...props
+  somethingToShow, meta: { focus }, execute, format, clippy, miri, assembly, llvmIr, mir, wasm, gist, ...props
 }) => {
   if (!somethingToShow) {
     return null;
@@ -139,6 +139,7 @@ const Output: React.SFC<OutputProps> = ({
         {focus === Focus.Execute && <SimplePane {...execute} kind="execute" />}
         {focus === Focus.Format && <Format {...format} />}
         {focus === Focus.Clippy && <SimplePane {...clippy} kind="clippy" />}
+        {focus === Focus.Miri && <SimplePane {...miri} kind="miri" />}
         {focus === Focus.Asm && <PaneWithCode {...assembly} kind="asm" />}
         {focus === Focus.LlvmIr && <PaneWithCode {...llvmIr} kind="llvm-ir" />}
         {focus === Focus.Mir && <PaneWithCode {...mir} kind="mir" />}
@@ -163,6 +164,10 @@ const Output: React.SFC<OutputProps> = ({
           label="Clippy"
           onClick={props.focusClippy}
           tabProps={clippy} />
+        <Tab kind={Focus.Miri} focus={focus}
+          label="Miri"
+          onClick={props.focusMiri}
+          tabProps={miri} />
         <Tab kind={Focus.Asm} focus={focus}
           label="ASM"
           onClick={props.focusAssembly}
@@ -208,6 +213,7 @@ interface OutputProps extends OutputState {
   focusExecute: () => void;
   focusFormat: () => void;
   focusClippy: () => void;
+  focusMiri: () => void;
   focusAssembly: () => void;
   focusLlvmIr: () => void;
   focusMir: () => void;
@@ -225,6 +231,7 @@ const mapDispatchToProps = ({
   focusExecute: () => changeFocus(Focus.Execute),
   focusFormat: () => changeFocus(Focus.Format),
   focusClippy: () => changeFocus(Focus.Clippy),
+  focusMiri: () => changeFocus(Focus.Miri),
   focusAssembly: () => changeFocus(Focus.Asm),
   focusLlvmIr: () => changeFocus(Focus.LlvmIr),
   focusMir: () => changeFocus(Focus.Mir),

--- a/ui/frontend/ToolsMenu.tsx
+++ b/ui/frontend/ToolsMenu.tsx
@@ -7,11 +7,13 @@ import MenuGroup from './MenuGroup';
 import {
   performClippy,
   performFormat,
+  performMiri,
 } from './actions';
 import State from './state';
 
 interface ToolsMenuProps {
   clippy: () => any;
+  miri: () => any;
   format: () => any;
   close: () => void;
 }
@@ -28,11 +30,18 @@ const ToolsMenu: React.SFC<ToolsMenuProps> = props => (
       onClick={() => { props.clippy(); props.close(); }}>
       Catch common mistakes and improve the code using the Clippy linter.
     </ButtonMenuItem>
+    <ButtonMenuItem
+      name="Miri"
+      onClick={() => { props.miri(); props.close(); }}>
+      Execute this program in the Miri interpreter to detect certain
+      cases of undefined behavior (like out-of-bounds memory access).
+    </ButtonMenuItem>
   </MenuGroup>
 );
 
 const mapDispatchToProps = ({
   clippy: performClippy,
+  miri: performMiri,
   format: performFormat,
 });
 

--- a/ui/frontend/actions.ts
+++ b/ui/frontend/actions.ts
@@ -23,6 +23,7 @@ const routes = {
   execute: { pathname: '/execute' },
   format: { pathname: '/format' },
   clippy: { pathname: '/clippy' },
+  miri: { pathname: '/miri' },
   meta: {
     crates: { pathname: '/meta/crates' },
     version: {
@@ -78,6 +79,9 @@ export enum ActionType {
   RequestClippy = 'REQUEST_CLIPPY',
   ClippySucceeded = 'CLIPPY_SUCCEEDED',
   ClippyFailed = 'CLIPPY_FAILED',
+  RequestMiri = 'REQUEST_MIRI',
+  MiriSucceeded = 'MIRI_SUCCEEDED',
+  MiriFailed = 'MIRI_FAILED',
   RequestGistLoad = 'REQUEST_GIST_LOAD',
   GistLoadSucceeded = 'GIST_LOAD_SUCCEEDED',
   GistLoadFailed = 'GIST_LOAD_FAILED',
@@ -387,6 +391,29 @@ export function performClippy(): ThunkAction {
   };
 }
 
+const requestMiri = () =>
+  createAction(ActionType.RequestMiri);
+
+const receiveMiriSuccess = ({ stdout, stderr }) =>
+  createAction(ActionType.MiriSucceeded, { stdout, stderr });
+
+const receiveMiriFailure = ({ error }) =>
+  createAction(ActionType.MiriFailed, { error });
+
+export function performMiri(): ThunkAction {
+  // TODO: Check a cache
+  return function(dispatch, getState) {
+    dispatch(requestMiri());
+
+    const { code } = getState();
+    const body = { code };
+
+    return jsonPost(routes.miri, body)
+      .then(json => dispatch(receiveMiriSuccess(json)))
+      .catch(json => dispatch(receiveMiriFailure(json)));
+  };
+}
+
 interface GistSuccessProps {
   id: string;
   url: string;
@@ -610,6 +637,9 @@ export type Action =
   | ReturnType<typeof requestClippy>
   | ReturnType<typeof receiveClippySuccess>
   | ReturnType<typeof receiveClippyFailure>
+  | ReturnType<typeof requestMiri>
+  | ReturnType<typeof receiveMiriSuccess>
+  | ReturnType<typeof receiveMiriFailure>
   | ReturnType<typeof requestGistLoad>
   | ReturnType<typeof receiveGistLoadSuccess>
   | ReturnType<typeof receiveGistLoadFailure>

--- a/ui/frontend/reducers/output/index.ts
+++ b/ui/frontend/reducers/output/index.ts
@@ -8,12 +8,14 @@ import gist from './gist';
 import llvmIr from './llvmIr';
 import meta from './meta';
 import mir from './mir';
+import miri from './miri';
 import wasm from './wasm';
 
 const output = combineReducers({
   meta,
   format,
   clippy,
+  miri,
   assembly,
   llvmIr,
   mir,

--- a/ui/frontend/reducers/output/meta.ts
+++ b/ui/frontend/reducers/output/meta.ts
@@ -17,6 +17,9 @@ export default function meta(state = DEFAULT, action: Action) {
     case ActionType.RequestClippy:
       return { ...state, focus: Focus.Clippy };
 
+    case ActionType.RequestMiri:
+      return { ...state, focus: Focus.Miri };
+
     case ActionType.CompileLlvmIrRequest:
       return { ...state, focus: Focus.LlvmIr };
 

--- a/ui/frontend/reducers/output/miri.ts
+++ b/ui/frontend/reducers/output/miri.ts
@@ -1,0 +1,31 @@
+import { Action, ActionType } from '../../actions';
+import { finish, start } from './sharedStateManagement';
+
+const DEFAULT: State = {
+  requestsInProgress: 0,
+  stdout: null,
+  stderr: null,
+  error: null,
+};
+
+interface State {
+  requestsInProgress: number;
+  stdout?: string;
+  stderr?: string;
+  error?: string;
+}
+
+export default function miri(state = DEFAULT, action: Action) {
+  switch (action.type) {
+    case ActionType.RequestMiri:
+      return start(DEFAULT, state);
+    case ActionType.MiriSucceeded: {
+      const { stdout = '', stderr = '' } = action;
+      return finish(state, { stdout, stderr });
+    }
+    case ActionType.MiriFailed:
+      return finish(state, { error: action.error });
+    default:
+      return state;
+  }
+}

--- a/ui/frontend/selectors/index.ts
+++ b/ui/frontend/selectors/index.ts
@@ -78,6 +78,7 @@ const getOutputs = (state: State) => [
   state.output.gist,
   state.output.llvmIr,
   state.output.mir,
+  state.output.miri,
   state.output.wasm,
 ];
 

--- a/ui/frontend/types.ts
+++ b/ui/frontend/types.ts
@@ -74,6 +74,7 @@ export enum Backtrace {
 
 export enum Focus {
   Clippy = 'clippy',
+  Miri = 'miri',
   LlvmIr = 'llvm-ir',
   Mir = 'mir',
   Wasm = 'wasm',

--- a/ui/src/main.rs
+++ b/ui/src/main.rs
@@ -99,6 +99,7 @@ fn main() {
     mount.mount("/execute", execute);
     mount.mount("/format", format);
     mount.mount("/clippy", clippy);
+    mount.mount("/miri", miri);
     mount.mount("/meta/crates", meta_crates);
     mount.mount("/meta/version/stable", meta_version_stable);
     mount.mount("/meta/version/beta", meta_version_beta);
@@ -190,6 +191,15 @@ fn clippy(req: &mut Request) -> IronResult<Response> {
         sandbox
             .clippy(&req.into())
             .map(ClippyResponse::from)
+            .map_err(Error::Sandbox)
+    })
+}
+
+fn miri(req: &mut Request) -> IronResult<Response> {
+    with_sandbox(req, |sandbox, req: MiriRequest| {
+        sandbox
+            .miri(&req.into())
+            .map(MiriResponse::from)
             .map_err(Error::Sandbox)
     })
 }
@@ -584,6 +594,18 @@ struct ClippyResponse {
     stderr: String,
 }
 
+#[derive(Debug, Clone, Deserialize)]
+struct MiriRequest {
+    code: String,
+}
+
+#[derive(Debug, Clone, Serialize)]
+struct MiriResponse {
+    success: bool,
+    stdout: String,
+    stderr: String,
+}
+
 #[derive(Debug, Clone, Serialize)]
 struct CrateInformation {
     name: String,
@@ -736,6 +758,24 @@ impl From<ClippyRequest> for sandbox::ClippyRequest {
 impl From<sandbox::ClippyResponse> for ClippyResponse {
     fn from(me: sandbox::ClippyResponse) -> Self {
         ClippyResponse {
+            success: me.success,
+            stdout: me.stdout,
+            stderr: me.stderr,
+        }
+    }
+}
+
+impl From<MiriRequest> for sandbox::MiriRequest {
+    fn from(me: MiriRequest) -> Self {
+        sandbox::MiriRequest {
+            code: me.code,
+        }
+    }
+}
+
+impl From<sandbox::MiriResponse> for MiriResponse {
+    fn from(me: sandbox::MiriResponse) -> Self {
+        MiriResponse {
             success: me.success,
             stdout: me.stdout,
             stderr: me.stderr,

--- a/ui/src/sandbox.rs
+++ b/ui/src/sandbox.rs
@@ -205,6 +205,19 @@ impl Sandbox {
         })
     }
 
+    pub fn miri(&self, req: &MiriRequest) -> Result<MiriResponse> {
+        self.write_source_code(&req.code)?;
+        let mut command = self.miri_command();
+
+        let output = command.output().map_err(Error::UnableToExecuteCompiler)?;
+
+        Ok(MiriResponse {
+            success: output.status.success(),
+            stdout: vec_to_str(output.stdout)?,
+            stderr: vec_to_str(output.stderr)?,
+        })
+    }
+
     pub fn crates(&self) -> Result<Vec<CrateInformation>> {
         let mut command = basic_secure_docker_command();
         command.args(&[Channel::Stable.container_name()]);
@@ -301,6 +314,16 @@ impl Sandbox {
         cmd.arg("clippy").args(&["cargo", "clippy"]);
 
         debug!("Clippy command is {:?}", cmd);
+
+        cmd
+    }
+
+    fn miri_command(&self) -> Command {
+        let mut cmd = self.docker_command(None);
+
+        cmd.arg("miri").args(&["cargo", "miri-playground"]);
+
+        debug!("Miri command is {:?}", cmd);
 
         cmd
     }
@@ -619,6 +642,18 @@ pub struct ClippyRequest {
 
 #[derive(Debug, Clone)]
 pub struct ClippyResponse {
+    pub success: bool,
+    pub stdout: String,
+    pub stderr: String,
+}
+
+#[derive(Debug, Clone)]
+pub struct MiriRequest {
+    pub code: String,
+}
+
+#[derive(Debug, Clone)]
+pub struct MiriResponse {
     pub success: bool,
     pub stdout: String,
     pub stderr: String,
@@ -965,6 +1000,28 @@ mod test {
 
         assert!(resp.stderr.contains("deny(eq_op)"));
         assert!(resp.stderr.contains("warn(zero_divided_by_zero)"));
+    }
+
+    #[test]
+    fn interpreting_code() -> Result<()> {
+        let code = r#"
+        fn main() {
+            let mut a: [u8; 0] = [];
+            unsafe { *a.get_unchecked_mut(1) = 1; }
+        }
+        "#;
+
+        let req = MiriRequest {
+            code: code.to_string(),
+        };
+
+        let sb = Sandbox::new()?;
+        let resp = sb.miri(&req)?;
+
+        assert!(resp.stderr.contains("pointer computed at offset 1"));
+        assert!(resp.stderr.contains("outside bounds of allocation"));
+        assert!(resp.stderr.contains("which has size 0"));
+        Ok(())
     }
 
     #[test]

